From 244d5ea359a0694d68d343525420ff09907c3a43 Mon Sep 17 00:00:00 2001
From: Noam Lessner <nlessner@habana.ai>
Date: Sun, 02 Feb 2020 11:50:19 +0200
Subject: [PATCH] [SW-8229] goya2: support image switching

Read flag in flash to allow switching between primary
and secondary image so we can switch between which image
partition to load first

Change-Id: Ifac01cd17967638488e920941af0fb08a212c741
Signed-off-by: Noam Lessner <nlessner@habana.ai>
---

diff --git a/app/arcboot_rom.c b/app/arcboot_rom.c
index 09cb8e3..07a7dba 100644
--- a/app/arcboot_rom.c
+++ b/app/arcboot_rom.c
@@ -83,8 +83,22 @@
 
 static uint32_t arcboot_flash_image_offset(uint8_t image_ptr)
 {
-	/* TODO: switch prim/secondary relative to flag in flash */
 	uint32_t flash_image_offset;
+	uint32_t flash_image_switch_offset;
+	uint8_t flash_image_switch;
+
+	/* read offset in flash to primary/secondary image selection */
+	flash_image_switch_offset =
+		spi_flash_readl(FLASH_IMAGE_SWITCH_OFFSET);
+
+	/* read if we need to switch order of primary/secondary image */
+	flash_image_switch = spi_flash_readl(flash_image_switch_offset);
+	if (flash_image_switch != 0) {
+		if (image_ptr != ARCBOOT_SECONDARY)
+			image_ptr = ARCBOOT_SECONDARY;
+		else
+			image_ptr = ARCBOOT_PRIMARY;
+	}
 
 	if (image_ptr == ARCBOOT_SECONDARY) {
 		flash_image_offset =
diff --git a/common/spi_flash.h b/common/spi_flash.h
index 555c2cc..5f1e690 100644
--- a/common/spi_flash.h
+++ b/common/spi_flash.h
@@ -41,9 +41,11 @@
 #define FLASH_REVOKE_BITS_PER_DWORD	4
 
 #define FLASH_PRIMARY_OFFSET		0
-#define FLASH_SECONDARY_OFFSET		4
-#define FLASH_EFUSE_OFFSET		8
-#define FLASH_IMAGE_PTR			12
+#define FLASH_SECONDARY_OFFSET		(FLASH_PRIMARY_OFFSET + \
+						sizeof(uint32_t))
+#define FLASH_EFUSE_OFFSET		(FLASH_SECONDARY_OFFSET + \
+						sizeof(uint32_t))
+#define FLASH_IMAGE_SWITCH_OFFSET	(FLASH_EFUSE_OFFSET + sizeof(uint32_t))
 
 #define FLASH_EFUSE_FW_VERSION_OFFSET	0
 #define FLASH_EFUSE_KEYS_OFFSET		16
